treeherder/model/derived/jobs.py

```diff
@@ -1212,16 +1229,41 @@ def _load_ref_and_job_data_structs(
 
         build_system_type = job.get('build_system_type', 'buildbot')
 
-        # Should be the buildername in the case of buildbot
+        sh = sha1()
+        sh.update(''.join(
+            map(lambda x: str(x),
+                [build_system_type, self.project, build_platform.os_name,
+                 build_platform.platform, build_platform.architecture,
+                 machine_platform.os_name, machine_platform.platform,
+                 machine_platform.architecture,
+                 job_group.name, job_group.symbol, job_type.name,
+                 job_type.symbol, option_collection_hash])))
+        signature_hash = sh.hexdigest()
+
+        # Should be the buildername in the case of buildbot (if not provided
+        # default to using the signature hash)
         reference_data_name = job.get('reference_data_name', None)
-
-        signature = self.refdata_model.add_reference_data_signature(
-            reference_data_name, build_system_type, self.project,
-            [build_system_type, self.project, build_os_name, build_platform, build_architecture,
-             machine_os_name, machine_platform, machine_architecture,
-             group_name, group_symbol, job_type, job_symbol,
-             option_collection_hash]
-        )
+        if not reference_data_name:
+            reference_data_name = signature_hash
+
+        (signature, _) = ReferenceDataSignatures.objects.get_or_create(
```

Comment:
ReferenceDataSignature has a unique_together index
```
unique_together = ('name', 'signature', 'build_system_type', 'repository')
```
so you should conditionally create the objects based on that, and put all the other fields in the defaults dict